chore: update browserlist and babel targets to chrome/firefox 89

davidmurdoch opened the pull request:
## **Description**

IIRC, we've documented extension browser support for versions 2 years old and newer (I can't find this documentation now though). Chrome and Firefox 89 were released in 2021, so the version in this PR should be good to go.

We could go with even newer browser versions, but it was a struggle getting it to work for v89, so I didn't try to take the versions any further.

This PR also reduces the chance of seeing a lavamoat caused `RegExpCompiler Allocation failed - process out of memory` error (https://github.com/MetaMask/metamask-extension/issues/21006).

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

davidmurdoch committed:
update browserlist and babel targets to chrome/firefox 89 to add support for top level async

davidmurdoch committed:
working around lavamoat issues consumes about 60% of my working life.

Gudahtt commented on .browserslistrc:
```diff
@@ -1 +1 @@
-chrome >= 80, firefox >= 78
+chrome >= 89, firefox >= 89
```
Looks like these versions are below the `OUTDATED_BROWSER_VERSIONS` we have configured, so this looks good :+1:

Gudahtt commented on .yarn/patches/lavamoat-core-npm-14.4.1-c4e8bbb016.patch:
```diff
@@ -1,31 +0,0 @@
-       // shows the full call stack
-       stackFiltering: 'verbose',
-+      // prevents most common override mistake cases from tripping up users
-+      overrideTaming: 'severe',
```
Is this patch still relevant to v15?

Gudahtt replied:
Ah, I see that this is line is in v15 already

Gudahtt commented on package.json:
```diff
@@ -189,7 +189,6 @@
     "@eslint/eslintrc@^2.0.1": "patch:@eslint/eslintrc@npm%3A2.0.2#./.yarn/patches/@eslint-eslintrc-npm-2.0.2-d308674d86.patch",
     "@fortawesome/fontawesome-free@^5.13.0": "patch:@fortawesome/fontawesome-free@npm%3A5.13.0#./.yarn/patches/@fortawesome-fontawesome-free-npm-5.13.0-f20fc0388d.patch",
     "@keystonehq/bc-ur-registry@^0.5.0-alpha.5": "patch:@keystonehq/bc-ur-registry@npm%3A0.5.0-alpha.5#./.yarn/patches/@keystonehq-bc-ur-registry-npm-0.5.0-alpha.5-b95c7992a6.patch",
-    "@lavamoat/lavapack@^3.1.0": "patch:@lavamoat/lavapack@npm%3A3.1.0#./.yarn/patches/@lavamoat-lavapack-npm-3.1.0-34c65d233b.patch",
```
Interesting, this patch was needed to remove absolute file paths from the bundle, but it looks like that was fixed upstream (relative file paths are used instead). Nice, one less patch needed!

Could you remove the patch file as well? It looks like it's still present in the `.yarn/patches` directory

Gudahtt replied:
I have just double checked to confirm that the absolute paths are gone. They're all relative paths now, so this patch is definitely safe to remove.

Gudahtt reviewed (changes requested):
Great work! We need to preserve one part of the `lavamoat-core` patch that was removed, and there is an extra patch file to delete for lavapack. Everything looks good aside from that.

davidmurdoch committed:
remove unused patch file

Gudahtt commented on babel.config.js:
```diff
@@ -5,8 +5,15 @@ module.exports = function (api) {
+      // logical assignment operators). This plugin lets us target es2020-level
+      // browsers (except we do still end up with transpiled logical assignment
+      // operators 😭)
+      '@babel/plugin-transform-logical-assignment-operators',
```
We'll need to list this in the depcheck config, as depcheck doesn't see where it's used

davidmurdoch committed:
skip policy write step to prevent intermittent build failures

davidmurdoch committed:
update lockfile

davidmurdoch committed:
add new package to .depcheck.yml

davidmurdoch commented on .yarn/patches/lavamoat-core-npm-14.4.1-c4e8bbb016.patch:
```diff
@@ -1,31 +0,0 @@
--    fs.writeFileSync(policyPath, jsonStringify(lavamoatPolicy, { space: 2 }))
-+    // Skip policy write step to prevent intermittent build failures
-+    // The extension validates the policy in a separate step, we don't need it
-+    // to be written to disk here.
```
Fixed! Good catch. I think I just ran into this bug a few minutes ago on another PR that depends on this one.

(Patching an existing yarn patch is absolute torture.)

Gudahtt reviewed (dismissed):
LGTM!

davidmurdoch committed:
lavamoat gunna lavamoat

davidmurdoch commented:
@Gudahtt I didn't see your review and comment till after I commit a fixed to the policies. Sorry!

dbrans commented:
> This PR also reduces the chance of seeing a lavamoat caused RegExpCompiler Allocation failed - process out of memory error (https://github.com/MetaMask/metamask-extension/issues/21006).

Out of curiosity, what is the connection between the browser version and the out of memory error?

davidmurdoch commented:
@dbrans @Gudahtt This already occurs at the current browser version levels, just not as frequently. Bumping the browser version seems to have caused additional regex memory usage.

I'm created a PR with the LavaMoat patches upstream: https://github.com/LavaMoat/LavaMoat/pull/867

Gudahtt commented:
@davidmurdoch is this ready to merge? I'm seeing the OOM issue on `develop` fairly often, so I'm eager to see if these patch solves the issue or not.

davidmurdoch commented:
@Gudahtt nothing blocking on my end. Hopefully the additions to the lavamoat patches will be short-lived!

legobeat reviewed (dismissed):
LGTM, though it needs a sync with `develop` and resulting conflict resolution before merging

legobeat committed:
Merge branch 'develop' into update-browser-targets

Gudahtt commented:
Looks like the policy update failed due to a syntax error in the policy file, which the update script isn't tolerant of. I'd recommend resetting the policies to match `develop`, then retrying the update-policies command

legobeat committed:
Merge branch 'develop' into update-browser-targets

legobeat committed:
Merge branch 'develop' into update-browser-targets

legobeat committed:
Merge branch 'develop' into update-browser-targets

Gudahtt commented:
Thanks, sounds good, I had thought the policy check was failing when I made that comment but I must have been looking at the wrong workflow.

Gudahtt commented on package.json:
```diff
@@ -218,7 +218,7 @@
     "nonce-tracker@npm:^3.0.0": "patch:nonce-tracker@npm%3A3.0.0#~/.yarn/patches/nonce-tracker-npm-3.0.0-c5e9a93f9d.patch",
     "@trezor/connect-web": "9.0.11",
-    "lavamoat-core@npm:^15.1.1": "patch:lavamoat-core@npm%3A15.1.1#~/.yarn/patches/lavamoat-core-npm-15.1.1-51fbe39988.patch"
+    "lavamoat-core@npm:^15.0.0": "patch:lavamoat-core@npm%3A15.0.0#~/.yarn/patches/lavamoat-core-npm-15.0.0-5f5fd568f9.patch"
```
It looks like this patch isn't getting applied (v15.1.1 is used in the lockfile). I think this explains the intermittent policy parse issues we're seeing. That intermittent policy parse issue is what the "Skip policy write step to prevent intermittent build failures" step of the patch is meant to fix

Gudahtt replied:
As an aside, this is not the first time I've noticed an unused patch go unnoticed. We should find a way to flag that in a CI check. We shouldn't have unused patches.

Gudahtt replied:
Oh, now the whole patch is gone. The regex optimizations need to be preserved

davidmurdoch replied:
Patching patch files is zero fun. Haha

legobeat committed:
fix lavamoat-core patch application

legobeat committed:
Merge branch 'develop' into update-browser-targets

legobeat committed:
update lavamoat build policy

legobeat committed:
Merge branch 'develop' into update-browser-targets

legobeat committed:
lavamoat-core patch: optimize regex allocation and compilation

port of https://github.com/LavaMoat/LavaMoat/pull/909
  from https://github.com/LavaMoat/LavaMoat/pull/867

legobeat committed:
lavamoat-core patch: add regex cache (#22607)

- Targeting: #22521 
- Lifted from: https://github.com/LavaMoat/LavaMoat/pull/876
  - https://github.com/LavaMoat/LavaMoat/pull/909

---

Co-Authored by: David Murdoch 187813+davidmurdoch@users.noreply.github.com

legobeat committed:
Merge branch 'develop' into update-browser-targets

Gudahtt reviewed (approved):
LGTM!
